src/ast/expr.rs

```diff
@@ -56,66 +58,93 @@ impl Display for Literal {
     }
 }
 
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
-pub enum Expr {
-    Lit(Literal),
-    Var(Symbol),
-    // TODO make this its own type
-    Call(Symbol, Vec<Self>),
+#[derive(Debug, Clone)]
+pub struct ResolvedVar {
+    pub name: Symbol,
+    pub sort: ArcSort,
 }
 
-#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
-pub enum NormExpr {
-    Call(Symbol, Vec<Symbol>),
+impl PartialEq for ResolvedVar {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.sort.name() == other.sort.name()
+    }
 }
 
-impl NormExpr {
-    pub fn to_expr(&self) -> Expr {
-        match self {
-            NormExpr::Call(op, args) => {
-                Expr::Call(*op, args.iter().map(|a| Expr::Var(*a)).collect())
-            }
-        }
+impl Eq for ResolvedVar {}
+
+impl Hash for ResolvedVar {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.sort.name().hash(state);
     }
+}
 
-    pub(crate) fn map_def_use(
-        &self,
-        fvar: &mut impl FnMut(Symbol, bool) -> Symbol,
-        is_def: bool,
-    ) -> NormExpr {
+impl SymbolLike for ResolvedVar {
+    fn to_symbol(&self) -> Symbol {
+        self.name
+    }
+}
+
+impl Display for ResolvedVar {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.name)
+    }
+}
+
+impl ToSexp for ResolvedVar {
+    fn to_sexp(&self) -> Sexp {
+        Sexp::Symbol(self.name.to_string())
+    }
+}
+
+pub type Expr = GenericExpr<Symbol, Symbol, ()>;
+pub(crate) type ResolvedExpr = GenericExpr<ResolvedCall, ResolvedVar, ()>;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone)]
+pub enum GenericExpr<Head, Leaf, Ann> {
```

Comment:
Add documentation explaining what a GenericExpr is, what the different type parameters mean, etc.

Also, consider giving this a shorter name (I don't think we need `Generic` in the name: anyone using it will see the type takes type parameters.). Maybe we can call this `Expr` and what is currently called `Expr` could be `RawExpr` or similar?